cmd/compute-domain-controller/computedomain.go

```diff
@@ -0,0 +1,297 @@
+	}
+	if len(cds) != 1 {
+		return nil, fmt.Errorf("multiple ComputeDomains with the same UID")
+	}
+	cd, ok := cds[0].(*nvapi.ComputeDomain)
+	if !ok {
+		return nil, fmt.Errorf("failed to cast to ComputeDomain")
+	}
+	return cd, nil
+}
+
+// RemoveFinalizer removes the finalizer from a ComputeDomain.
+func (m *ComputeDomainManager) RemoveFinalizer(ctx context.Context, uid string) error {
+	cd, err := m.Get(uid)
+	if err != nil {
+		return fmt.Errorf("error retrieving ComputeDomain: %w", err)
+	}
+	if cd == nil {
+		return nil
+	}
+
+	if cd.GetDeletionTimestamp() == nil {
+		return fmt.Errorf("attempting to remove finalizer before ComputeDomain marked for deletion")
+	}
+
+	newCD := cd.DeepCopy()
+	newCD.Finalizers = []string{}
+	for _, f := range cd.Finalizers {
+		if f != computeDomainFinalizer {
+			newCD.Finalizers = append(newCD.Finalizers, f)
+		}
+	}
+	if len(cd.Finalizers) == len(newCD.Finalizers) {
+		return nil
+	}
+
+	if _, err = m.config.clientsets.Nvidia.ResourceV1beta1().ComputeDomains(cd.Namespace).Update(ctx, newCD, metav1.UpdateOptions{}); err != nil {
+		return fmt.Errorf("error updating ComputeDomain: %w", err)
+	}
+
+	return nil
+}
+
+// AssertWorkloadsCompletes ensures that all workloads asssociated with a ComputeDomain have completed.
+//
+// TODO: We should probably also check to ensure that all ResourceClaims
+// generated from our ResourceClaimTemplate for workloads are gone. Doing
+// something is better than nothing for now though.
+func (m *ComputeDomainManager) AssertWorkloadsCompleted(ctx context.Context, cdUID string) error {
+	labelSelector := &metav1.LabelSelector{
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      computeDomainLabelKey,
+				Operator: metav1.LabelSelectorOpIn,
+				Values:   []string{cdUID},
+			},
+		},
+	}
+
+	nodes, err := m.config.clientsets.Core.CoreV1().Nodes().List(ctx, metav1.ListOptions{
+		LabelSelector: metav1.FormatLabelSelector(labelSelector),
+	})
+	if err != nil {
+		return fmt.Errorf("error retrieving nodes: %w", err)
+	}
+
+	if len(nodes.Items) != 0 {
+		return fmt.Errorf("nodes exist with label for ComputeDomain %s", cdUID)
+	}
+
+	return nil
+}
+
+func (m *ComputeDomainManager) addFinalizer(ctx context.Context, cd *nvapi.ComputeDomain) error {
+	for _, f := range cd.Finalizers {
+		if f == computeDomainFinalizer {
+			return nil
+		}
+	}
+
+	newCD := cd.DeepCopy()
+	newCD.Finalizers = append(newCD.Finalizers, computeDomainFinalizer)
+	if _, err := m.config.clientsets.Nvidia.ResourceV1beta1().ComputeDomains(cd.Namespace).Update(ctx, newCD, metav1.UpdateOptions{}); err != nil {
+		return fmt.Errorf("error updating ComputeDomain: %w", err)
+	}
+
+	return nil
+}
+
+func (m *ComputeDomainManager) onAddOrUpdate(ctx context.Context, obj any) error {
+	cd, ok := obj.(*nvapi.ComputeDomain)
+	if !ok {
+		return fmt.Errorf("failed to cast to ComputeDomain")
+	}
+
+	klog.Infof("Processing added or updated ComputeDomain: %s/%s", cd.Namespace, cd.Name)
```

Comment:
We talked about invoking the same callback in both cases.

Now reading this log msg I wonder: for debugging, will it ever be relevant to distinguish both cases?

Reply:
Something to consider, but maybe that is better handled in the wrapping anonymous function where these are registered so we can continue to reuse the core logic.